### ISSUE TITLE
SerializationException on LogContext when using xUnit.net

### ISSUE THREAD

**Issue by christophdebaene** [serilog/serilog#109]
Creating a simple class library with ````xunit```` with the following test doesn't work when using ````LogContext````

````C#
public class LogContextTests
{
   [Fact]
   public void LogContextSample()
   {         
      var log = new LoggerConfiguration()
         .Enrich.FromLogContext()                
         .CreateLogger();

         using (LogContext.PushProperty("A", 1))
         {
         }            
   }
}
````
With the console version ````xunit.console.clr4.exe```` [xUnit.net: Runners] (https://www.nuget.org/packages/xunit.runners) it shows the following exception

````
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.Runtime.Serialization.SerializationException: Unable to find assembly 'Serilog.FullNetFx, Version=1.2.0.0, Culture=neutral, PublicKeyToken=24c2f752a8e58a10'.

Server stack trace: 
   at System.Runtime.Serialization.Formatters.Binary.BinaryAssemblyInfo.GetAssembly()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.GetType(BinaryAssemblyInfo assemblyInfo, String name)
   at System.Runtime.Serialization.Formatters.Binary.ObjectMap..ctor(String objectName, String[] memberNames, BinaryTypeEnum[] binaryTypeEnumA, Object[] typeInformationA, Int32[] memberAssemIds, ObjectReader objectReader, Int32 objectId, BinaryAssemblyInfo assemblyInfo, SizedArray assemIdToAssemblyTable)
   at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.ReadObjectWithMapTyped(BinaryObjectWithMapTyped record)
   at System.Runtime.Serialization.Formatters.Binary.__BinaryParser.Run()
   at System.Runtime.Serialization.Formatters.Binary.ObjectReader.Deserialize(HeaderHandler handler, __BinaryParser serParser, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Serialization.Formatters.Binary.BinaryFormatter.Deserialize(Stream serializationStream, HeaderHandler handler, Boolean fCheck, Boolean isCrossAppDomain, IMethodCallMessage methodCallMessage)
   at System.Runtime.Remoting.Channels.CrossAppDomainSerializer.DeserializeObject(MemoryStream stm)
   at System.Runtime.Remoting.Messaging.SmuggledMethodCallMessage.FixupForNewAppDomain()
   at System.Runtime.Remoting.Channels.CrossAppDomainSink.DoDispatch(Byte[] reqStmBuff, SmuggledMethodCallMessage smuggledMcm, SmuggledMethodReturnMessage& smuggledMrm)
   at System.Runtime.Remoting.Channels.CrossAppDomainSink.DoTransitionDispatchCallback(Object[] args)

Exception rethrown at [0]: 
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at System.Web.UI.ICallbackEventHandler.RaiseCallbackEvent(String eventArgument)
   --- End of inner exception stack trace ---
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] parameters, CultureInfo culture)
   at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
   at Xunit.Sdk.Executor.OnTestResult(ITestResult result, ExecutorCallback callback)
   at Xunit.Sdk.TestClassCommandRunner.Execute(ITestClassCommand testClassCommand, List`1 methods, Predicate`1 startCallback, Predicate`1 resultCallback)
   at Xunit.Sdk.Executor.ThreadRunner(Object threadStart)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
   at System.Threading.ThreadHelper.ThreadStart(Object obj)
````
I think it's related with the following issue http://xunit.codeplex.com/workitem/9760.
Maybe it's only related to xUnit, but I want to be sure. Maybe the data is not serializable? 
It would be nice that all features of ````Serilog```` can be used through xUnit.net.

**Comment by plioi** [comment]
Data that is being placed in the logical call contexts should be serializable.  http://msdn.microsoft.com/en-us/library/system.runtime.remoting.messaging.callcontext.logicalsetdata%28v=vs.110%29.aspx .

If the things serilog puts there aren't serialization friendly, could this issue be reopened?  This affects at least one other test framework (Fixie).

**Comment by nblumhardt** [comment]
@plioi thanks for the follow-up. If this could be done smoothly I'd definitely be supportive of the change - though serializability alone won't ensure required DLLs are available in the target (remote) app domain, which then still fails on deserialization.

(Remoting's just horribly broken in this scenario unfortunately :-( ... Please do send a PR if you manage to unravel this one :-) )

**Comment by DmitryNaumov** [comment]
Just ran into same issue :(
But I've got an idea! What if using remoting surrogate selector or implementing ISerializable make a fake serialization, so nothing from Serilog.* assembly would be sent to remote side. What do you think? If it make sense I can do coding and submit PR.

**Comment by nblumhardt** [comment]
@DmitryNaumov would love to see if you can come up with something. I'll reopen in the hope we can make some progress here :+1:

**Comment by DmitryNaumov** [comment]
@nblumhardt I'll try to come up with PR

Meanwhile looks like
`using (LogContext.Suspend()) {...}`

is better than
`CallContext.FreeNamedDataSlot(typeof(LogContext).FullName);`

**Comment by nblumhardt** [comment]
@DmitryNaumov thanks for the update. Glad to hear `PermitCrossAppDomainCalls` has this covered - I had a recollection that it wasn't enough last time I investigated, but it would be great if it is.

I would be satisfied recommending `PermitCrossAppDomainCalls` for test projects if it works. I'm uneasy about enabling it by default - it's not a very well-tested path - but on the plus side it would be nice to have it "just work" in more cases.

Getting some feedback on whether it solves the issue for all cases (and thus isn't dependent on any GAC/path coincidences) would be a great first step. @plioi @christophdebaene ?

**Comment by plioi** [comment]
FWIW, my repo case using Fixie only passes when I call `CallContext.FreeNamedDataSlot(typeof(LogContext).FullName);` at the end of a test method.

If I instead wrap my test method content in `using (LogContext.Suspend()) {...}`, I get the exception.

if I instead set `LogContext.PermitCrossAppDomainCalls = false;` at the start of a test method, I get the exception.

**Comment by plioi** [comment]
And, surprisingly, I can no longer reproduce this with xUnit.

**Comment by plioi** [comment]
Also works under the latest NUnit.  I could use some guidance if this is the test runner's issue to resolve! I can't see what those test frameworks do that could affect this.

**Comment by nblumhardt** [comment]
Thanks for the interesting data @plioi. Are you using Serilog 1.5 or the new beta?

(Rather than call `Suspend()` in a `using` block, the idea would be to call `Suspend()` at the end of the test method and never `Dispose()` the returned object, I think. Even better if it's fixed entirely :-))

**Comment by ivog** [comment]
I'm also running into a SerializationException when using PushProperty in combination with nhibernate. It doesn't matter if i use nunit or xunit, both give me the exception. This is my code:

```
        [Fact]
        public void test()
        {
            LogContext.PushProperty("test", "bla");
            Log.Debug("test");

            ISessionFactory factory = FluentNHibernate.Cfg.Fluently.Configure()
                .Mappings(x => x.FluentMappings.AddFromAssemblyOf<Class1>())
                .Database(SQLiteConfiguration.Standard.UsingFile("firstProject.db"))
                .BuildSessionFactory();
        }
```
or you can find the repro code here: https://github.com/ivog/serilog_problem_repro.git
I've also saw this same exception in combination with Spring.Data. Hopefully I can find some time to dive deeper into this problem.

**Comment by ivog** [comment]
By the way, I can run the above code by christophdebaene without any problems.

**Comment by DmitryNaumov** [comment]
@ivog did you try using Suspend or PermitCrossAppDomainCalls?

**Comment by ivog** [comment]
I'm not entirely sure how to use those correctly but when i use the below code I get the same exception:

```
        [Test]
        public void TestWithNunit()
        {
            LogContext.PermitCrossAppDomainCalls = false;

            using (var context = LogContext.Suspend())
            {
                LogContext.PushProperty("test", "bla");
                Log.Debug("test");

                ISessionFactory factory = FluentNHibernate.Cfg.Fluently.Configure()
                    .Mappings(x => x.FluentMappings.AddFromAssemblyOf<Class1>())
                    .Database(SQLiteConfiguration.Standard.UsingFile("firstProject.db"))
                    .BuildSessionFactory();
            }
        }
```

**Comment by DmitryNaumov** [comment]
@ivog I don't understand how building NH's session factory involves remoting calls, but I missed your link with solution, so now I try to repro it on my local machine.

**Comment by ivog** [comment]
Thanks Dmitry, I haven't been able to dive further into this issue yet, am curious if you're able to reproduce it.

**Comment by DmitryNaumov** [comment]
@ivog I ran your repro test and found no problems. As I said before with some knowledge of NH's nuts and bolts, I don't understand why SerializationException may be thrown. The only thing I can imagine is that it is not NH-related, but somewhere you have special settings for xUnit/NUnit, which tells test runner to run tests in separate domain. If so, please tell me where I can configure it as well.

**Comment by DmitryNaumov** [comment]
![repro](https://cloud.githubusercontent.com/assets/866389/13437656/6dad424e-dff6-11e5-8537-d1a850ec39df.jpg)

Meanwhile I managed to repro some kind of problem using VS 2015 "Profile test" feature. It does exactly what I need - runs test in separate AppDomain.

**Comment by DmitryNaumov** [comment]
@ivog This usage of LogContext.Suspend helps

```
        [Fact]
        public void test()
        {
            LogContext.PushProperty("test", "bla");

            using (LogContext.Suspend())
            {
                Log.Debug("test");

                ISessionFactory factory = FluentNHibernate.Cfg.Fluently.Configure()
                    .Mappings(x => x.FluentMappings.AddFromAssemblyOf<Class1>())
                    .Database(SQLiteConfiguration.Standard.UsingFile("firstProject.db"))
                    .ExposeConfiguration(BuildSchema)
                    .BuildSessionFactory();

                factory.OpenSession();
            }
        }
```

@nblumhardt looks like I understand the problem: setting PermitCrossAppDomainCalls  doesn't clear existing CallContext slot, neither prevents setting it later on before remoting call. Same thing with Suspend - it allows inproper use like this (setting context property while Suspend is "active"):

```
            using (var context = LogContext.Suspend())
            {
                LogContext.PushProperty("test", "bla");
```

Let me think a while and I'll come up with PR

**Comment by ivog** [comment]
@DmitryNaumov I'm not aware of having any special test settings configured. Normally I run my tests using resharper with 'xunit test support' extension. I've also tried running the same test with xunit console runner and nunit console runner. Both give me the same exception. What testrunner are you using?

![image](https://cloud.githubusercontent.com/assets/2332637/13443462/74a2e5b2-e001-11e5-813f-0b7ec881ed04.png)

![image](https://cloud.githubusercontent.com/assets/2332637/13443447/63d6a6ce-e001-11e5-98b7-30fcfa1a03fd.png)

**Comment by ivog** [comment]
LogContext.Suspend indeed helps, but I don't think that will help me out. I'm trying to use PushProperty to capture logging of my applicationcode in my test output as described here: http://dhickey.ie/2015/06/capturing-log-output-in-tests-with-xunit2/

**Comment by plioi** [comment]
@nblumhardt My apologies for the delay to your request. Here's my testing scenarios, for the original repro steps at the top of this page, on various test frameworks. I used the latest stable version of Serilog as well as each test framework:

Serilog 1.5.14
NUnit 3.0.1 - passes
xUnit 2.1.0 - passes
Fixie 1.0.0.41 - fails with the exception from the original repro, "SerializationException: Unable to find assembly 'Serilog.FullNetFx, Version=1.5.0.0..."

First, is it surprising that the exception message is talking about **1.5.0.0**, while the actual assembly's "File version" is 1.5.14.370 and "Product version" is 1.5.14?

Second, I can make the Fixie scenario pass if I change it in any of the following ways:

1. Place `CallContext.FreeNamedDataSlot(typeof(LogContext).FullName);` as the last line of the test.
2. Place `LogContext.Suspend();` as the last line of the test.

I do **not** have success in Fixie when I instead try `LogContext.PermitCrossAppDomainCalls = false;` near the start of the tests.

Given that xUnit and NUnit work without intervention, there's got to be something that they're doing to be "better citizens" with this sort of use case. If you had to guess, what kind of cleanup would you *want* a test framework to be performing so that their use of AppDomains doesn't give Serilog trouble at test time? If I can find what they're doing, I can replicate it.

**Comment by plioi** [comment]
FWIW, here's a clue.  Fusion Log Viewer shows it's attempting to load Serilog.FullNetFx.DLL from the wrong folder: the folder the test runner lives in, despite the fact that Fixie does set the right application base path in the test running AppDomain.

**Comment by nblumhardt** [comment]
@plioi the assembly version is fixed at 1.5.0.0, so nothing unexpected here.

I think the `FreeNamedDataSlot()` approach is the way to go- another report here with verified "fix" https://github.com/serilog/serilog/issues/712#issuecomment-206947142

Seems like it should be quite a straightforward change to make `LogContext` free the slot when there's no data left, rather than leave an empty stack on it. Help appreciated! Will be great to finally put this one to bed.

**Comment by DmitryNaumov** [comment]
@nblumhardt Unfortunately you don't always control when remoting happens, that's my case. So you can't put FreeNamedDataSlot right before the call.

The problem with LogContext.Suspend, as far as I remember, is that after being "suspended" it clears CallContext (via setting Enrichers = null), but if someone on execution path will do, for example, PushProperty, CallContext's data slot will be filled again. And back to square 1, we have same problem.

I've been quite busy last month, but now I'm ready to help, We just need to decide which way to go.

**Comment by DmitryNaumov** [comment]
My suggestion is: when LogContext is "suspended" then
- Clear CallContext's data slot
- Do not allow to set it, until LogContext "resumed"

**Comment by nblumhardt** [comment]
@DmitryNaumov I think that would be a useful tweak to `Suspend()`, it sounds interesting, anyway.

In most of the examples so far though, the log context _is_ empty - so the change to the "popping" behaivour should fix a large number of these without needing to tinker with `Suspend()`.

Interestingly, if we make the first change, then `Suspend()` will end up clearing the data slot because it sets the context to an empty stack. Disallowing further "pushes" is an extension we could subsequently consider. I'd rather get the change to _pop_ stabilised first though - one thing at a time :-)

Thanks for the weigh-in!

**Comment by DmitryNaumov** [comment]
I went the way of `FreeNamedDataSlot()`, which seems to be simple:

```
        public static IDisposable Suspend()
        {
            var stack = GetOrCreateEnricherStack();
            var bookmark = new ContextStackBookmark(stack);

            //Enrichers = null;
            CallContext.FreeNamedDataSlot(DataSlotName);

            return bookmark;
        }

```
But one thing which bothers me is that it changes the behavior described in `Suspend()` comments:
```
        /// <summary>
        /// Remove all data from the context so that
        /// cross-<see cref="AppDomain"/> calls can be made without requiring
        /// Serilog assemblies to be present in the remote domain.
        /// </summary>
```

Actually now it doesn't remove data and at least one test (WhenSuspendedAllPropertiesAreRemovedFromTheContext) should be rewritten, because now if you log something during 'suspend', context properties will be still included in log event. The only difference is that they won't cross AppDomain boundary.

@nblumhardt Are you ok with such change in behavior or it may have consequences I can't foresee? Or, `Suspend()` should be renamed/moved to other place because it doesn't affects LogContext itself and only changes remoting behavior.

**Comment by nblumhardt** [comment]
@DmitryNaumov this doesn't sound quite right - during Suspend, if it's implemented as in your example above, no properties should be attached... Is it possible the test is running with the `dotnet5.1` Serilog build, which uses thread-local storage for LogContext and not the data slot?

Thanks for pushing forward!

**Comment by DmitryNaumov** [comment]
@nblumhardt let me rephrase my question - let's say I know how to avoid problems with remoting without actually clearing LogContext, then should `Suspend()` clear it for other reasons I can't foresee?

**Comment by DmitryNaumov** [comment]
Another problem which is not addressed in current code, imo, is that after `Suspend()` somewhere below in execution path you can 'start' new stack by using `PushProperty()` or `Enrich()` which will lead to same problems that we trying to solve by calling `Suspend()`

**Comment by DmitryNaumov** [comment]
Finally I did it (#733), but few things left open:
- Do we need `PermitCrossAppDomainCalls` and, if yes, what is scenario.
- Now `LogContext.Suspend()` has no relation to remoting, it's XML comments should be cleared out and the only purpose of it's existence is to clear current stack of enrichers to start new one. I don't know who needs it.

**Comment by nblumhardt** [comment]
Circling back on this to answer your questions made me think of a snippet of code I saw yesterday. Looks like `ObjectHandle` is what we've needed all along:

https://msdn.microsoft.com/en-us/library/system.runtime.remoting.objecthandle(v=vs.110).aspx

This is the usage, inherited from Microsoft:

https://github.com/serilog/serilog-extensions-logging/blob/dev/src/Serilog.Extensions.Logging/Extensions/Logging/SerilogLoggerProvider.cs#L70

This should allow us to scrap `PermitCrossAppDomainCalls` and our own `Wrapper` class entirely, which I'd love to do for 2.0.

The questionable semantics of `Suspend` mean we should probably just remove it.

What do you think @DmitryNaumov ?

**Comment by DmitryNaumov** [comment]
@nblumhardt Brilliant! This is the thing I was looking for, but my memory didn't give me this answer. I've double checked and it works like a charm - the metadata (assembly) is loaded only if someone in another domain tries to call ObjectHandle.Unwrap(). In our case this would mean "other domain uses serilog and it's fine and safe to access serilog types".

I'm two hands up to remove both PermitCrossAppDomainCalls as well as misleading Suspend().
Let me do it in code, but this will affect a lot of tests, due to Suspend() going away...

**Comment by nblumhardt** [comment]
@DmitryNaumov excellent news! :-)

**Comment by DmitryNaumov** [comment]
@nblumhardt been busy, but will complete it till the end of the week.
One thing that bothers me and I mentioned it before - when using ObjectHandle the data in CallContext is transfered via remoting no matter is it need it or not on remote side. May be I the last man using remoting and when I talk about remoting I mean not appdomains or processes on same host, but remote hosts, connected by wire. And on the wire, imo, every byte counts. So I can't get my hands into it before we decide - do we keep PermitCrossAppDomainCalls for such cases when we want to prevent any extra byte to be transferred to remote side or I'm overcautious and it doesn't make sense.

**Comment by nblumhardt** [comment]
@DmitryNaumov that's a really good question. Why is there always some "gotcha" lurking in this feature?! :)

I don't like `PermitCrossAppDomainCalls` much as a solution in this scenario; perhaps making `LogContext.Suspend()` completely clear the `CallContext` slot is the right way to cover this?

**Comment by DmitryNaumov** [comment]
@nblumhardt deeper in the woods lives bigger wolves.
And my vote is against Suspend :) Because it clears everything on caller's side. Mean after Suspend (just to do remoting call) logging context is gone, if I log that remoting call was successful, LogEvent won't be enriched with needful contextual information which was in LogContext. So my order of preference is:
1. Leave PermitCrossAppDomainCalls, remove Suspend
2. Forget about remoting, it is history. Remove both PCADC and Suspend
3. Alternate solutions...

**Comment by nblumhardt** [comment]
@DmitryNaumov `Suspend()` restores the `LogContext

**Comment by nblumhardt** [comment]
(Sorry, clicked that Comment button accidentally :-) )

**Comment by DmitryNaumov** [comment]
@nblumhardt yes, but let me make thing bit more complicated :) We use some 3rd party library. When some it's methods called, it does some remoting calls. So I can wrap this library with kinda facade and each call wrap with Suspend... but library raises events. Our event handlers do some logging and I need LogContext there :( So, Suspend wasn't an option for me from the beginning of this discussion. Carthage must be destroyed :) So imo the question is not about Suspend, it should be removed, but only about PCADC. Let's roll the dice to decide its future ? :)

**Comment by nblumhardt** [comment]
Yikes! :-)

It's starting to sound, to me, like accepting the few extra bytes for the objectref when using remoting for RPC might be the way to go.

It would mean we can axe PCADC, and (unless anyone has a use case for it?) Suspend() as well. 2.0 could kick off with a clean slate, and any scenarios left over can be dealt with carefully, knowing everything we now know....?

**Comment by DmitryNaumov** [comment]
It will be few extra bytes, if whatever we store in CallContext is MBRO, or it can be much more, if it is Serializable. In our case, at first glance, ImmutableStack<T>, which is stored in CallContext, is Serializable. That's why I keep talking about those who still uses remoting :-) and who may suffer from a lot of extra bytes transferred by wire.

So here is my two-steps plan:
1. Remove Suspend, redundant tests etc. but leave PCADC as is.
2. If it will still look ugly, remove PCADC as well

Agree?

**Comment by nblumhardt** [comment]
`ImmutableStack` is only serializable to support PCADC :-) .. So if we move to `ObjectWrapper`, we can just remove `[Serializable]`/`ISerializable` from `ImmutableStack`, kill PCADC and `Suspend()` and we should be in just-a-few-bytes territory.

**Comment by DmitryNaumov** [comment]
Do you mean remove [Serializable]/ISerializable and do not inherit from MBRO? So it means we never going to transfer our current stack of enrichers to another domain. Even in case (sorry for being so creative in imagining use cases which may never exists, but this one is close to my) if someone has process with main AppDomain, where logging is set up, it receives some external requests (now not remoting :-) but HTTP) and RequestId is added to LogContext via enricher, but delegates processing to other AppDomain (which behaves as sandbox). Obviously someone may want to log some info from that second domain and it would be nice if the logged information will be enriched is RequestId and other things which are added in main domain...

**Comment by nblumhardt** [comment]
Yes, that's what I mean :+1: 

Your scenario's not so far-fetched, we do pretty much the same thing in Seq internally (Seq Apps are appdomain-isolated). At that point, however, you hit the problem of needing to marshal a logger through, which is where things get tricky anyway.

In Seq there's a custom logging interface exposed via MBRO into the sandboxed appdomain, allowing the sandboxed code to write log messages, but these aren't pushed through the logger until they reach the parent appdomain.

Not sure if `LogContext` matters much in my case, but once you're logging from child appdomains you have to deal with this sort of plumbing anyway. (A cross-appdomain logging bridge would be a cool library for us to write/publish at some point! :-))

Given the major problem this PR's addressing is the widespread unit testing scenario, I'm still convinced we should use objectrefs and be done with it. Seems to be the option with the least complexity where all the other scenarios will "just work" - are there any other drawbacks apart from the transfer of the ref when doing RPC?

Cheers!

**Comment by DmitryNaumov** [comment]
@nblumhardt I have no questions left. Ready to start!

**Comment by nblumhardt** [comment]
Great! :-) :+1:

**Comment by DmitryNaumov** [comment]
It took me a while to understand how project things changed since moving to ASP.NET Core RC2. But now I think I've done. The most important test (DoesNotPreventCrossDomainCalls) skipped due to problems with dontext xunit console runner, but I've run it locally which helped me to figure out problem with slot names.
